### ISSUE TITLE
Lambda-binning with consistent treatment of boundaries

### ISSUE THREAD

**Issue by hdrake** [NOAA-GFDL/xwmt#39]

**Comment by hdrake** [comment]
> While the vertical transform/regridding is conservative, the mismatch between the user-input lambda bins and the vertical diagnostic grid results in compensating spurious positive/negative water mass transformations (which have no net effect).

Actually, these errors are not even always compensating, and can just straight up bias the results. E.g.: https://github.com/hdrake/xwmt/blob/53d9a8afda70993f082ec59a00aff8918f74dd40/notebooks/testing_examples.ipynb

**Comment by gmacgilchrist** [comment]
This is a very useful test case.

I'm still trying to get my head fully around how the `extrapolate` (or `extend`) could be the source of this error. Is it best understood as the fact that the extrapolation "fills out" a `lambda` bin that, in the model, would not have been filled out? I wonder what would be the more appropriate strategy to do this in a manner consistent with what's happening in the model. Online, _only_ the `lambda`s on the tracer points "exists", so any method to fill out the bins is going to introduce an error.

I saw a lot of issues at the `lamda` boundary when working through idealized cases before. See (the rather messy notebook) [here](https://github.com/gmacgilchrist/wmt_bgc/blob/master/notebooks/explore_box_wmt.ipynb) - the noise emerges in the WMT term when using either `xhistogram` or conservatively remapped coordinates (this was a precursor to `xgcm.transform`). Do you think this is the same thing, @hdrake (appreciate that this notebook may be a little tricky to interpret)? We suspected it was something to do with the discretization, but had not pinned down where.

**Comment by hdrake** [comment]
> I'm still trying to get my head fully around how the `extrapolate` (or `extend`) could be the source of this error. Is it best understood as the fact that the extrapolation "fills out" a `lambda` bin that, in the model, would not have been filled out?

Yes, I think that's essentially the issue, but I am also still exploring it.

**Comment by hdrake** [comment]
A related problem is how to deal with formal discontinuities in the watermass transformation framework when the vertical coordinate is not strictly monotonic, which has two distinct parts to it: 1) vanishing lambda gradients and 2) allowing lambda(r) to have multiple values means that information is lost in the binning process.

See the results of the following idealized experiments (w/ analytical solutions for contexts): https://github.com/hdrake/xwmt/blob/f9cb2bc32490b022485a80a60c08c681efea48d7/notebooks/testing_examples.ipynb

The analytical solutions show that the watermass transformations tend towards ±infinity as gradients go to zero. In the discrete world, these manifest as water mass transformation rates of very larger magnitude. It seems as though this can be particularly problematic if the tendency terms also change sign across the region of vanishing gradient, since then there is a discontinuity from a large positive water mass transformation to a large negative water mass transformation.

![Screenshot 2023-05-01 at 1 52 41 PM](https://user-images.githubusercontent.com/12971166/235500780-6615d22f-8820-4c52-ae12-f69f2c242b38.png)

**Comment by hdrake** [comment]
Upon further examination (see [notebook is most recent commit](https://github.com/hdrake/xwmt/blob/f9cb2bc32490b022485a80a60c08c681efea48d7/notebooks/testing_examples.ipynb)), I actually think the lambda-binning behavior is consistent as long as you think of the resulting water mass transformations as averages over the finite layers that they represent.

The plots in the notebooks above are misleading because I am comparing *bin-averaged* values from the discretized `xwmt` calculation against local values from the analytical solution. Since the analytical solutions are strongly skewed near the singularities, their local values at the mid-point of a lambda-bin are much lower than the average over the bin.

![Screenshot 2023-05-01 at 2 38 29 PM](https://user-images.githubusercontent.com/12971166/235509212-0d878d22-e7aa-4e6e-9135-288b7f94dc21.png)

Somewhat surprising, these actually still remain rather well behaved when you approach vanishing gradients!

**Comment by hdrake** [comment]
Given the above https://github.com/NOAA-GFDL/xwmt/issues/39#issuecomment-1530063105, I don't think there is actually a problem with how we are binning. However, I think we should leave this open until we have explored this a bit further and until I have turned these comparisons to analytical solutions to quantitative tests for `xwmt`.

**Comment by hdrake** [comment]
`pytest` suite that checks convergence with analytical solutions is now implemented in https://github.com/hdrake/xwmt/commit/e1e0f830994dcf15f504dfea5dbdb48a93e06a46
